File: src/config-command.ts
import { spawn } from "node:child_process";
import { configPath, ensureConfigFile } from "./config.js";
import { parseShellWords, runLocal } from "./ssh.js";

export interface ConfigCommandOptions {
  path?: boolean;
  editor?: string;
}

export async function runConfigCommand(options: ConfigCommandOptions): Promise<void> {
  if (options.path) {
    console.log(configPath());
    return;
  }

  const path = await ensureConfigFile();
  const editorCommand = options.editor ?? process.env.VISUAL ?? process.env.EDITOR;
  if (editorCommand && editorCommand.trim().length > 0) {
    await runEditor(editorCommand, path);
    return;
  }

  if (process.platform === "darwin") {
    const result = await runLocal("open", ["-t", path]);
    if (result.exitCode !== 0) throw new Error(result.stderr.trim() || `Failed to open ${path}`);
    console.log(`opened ${path}`);
    return;
  }

  console.log(path);
}

async function runEditor(editorCommand: string, path: string): Promise<void> {
  const parts = parseShellWords(editorCommand);
  if (parts.length === 0) throw new Error("Editor command is empty");
  await new Promise<void>((resolve, reject) => {
    const child = spawn(parts[0], [...parts.slice(1), path], { stdio: "inherit" });
    child.on("error", reject);
    child.on("close", (code) => {
      if (code === 0) resolve();
      else reject(new Error(`Editor exited with code ${code ?? 1}`));
    });
  });
}

File: src-rs/config_command.rs
use anyhow::{Result, bail};

use crate::config::{config_path, ensure_config_file};
use crate::ssh::{parse_shell_words, run_local, run_local_inherit};

pub fn run_config_command(print_path: bool, editor: Option<&str>) -> Result<()> {
    if print_path {
        println!("{}", config_path().display());
        return Ok(());
    }

    let path = ensure_config_file()?;
    let path_string = path.display().to_string();
    let editor_command = editor
        .map(str::to_string)
        .or_else(|| std::env::var("VISUAL").ok())
        .or_else(|| std::env::var("EDITOR").ok());

    if let Some(editor_command) = editor_command.filter(|value| !value.trim().is_empty()) {
        let parts = parse_shell_words(&editor_command)?;
        let Some((program, args)) = parts.split_first() else {
            bail!("editor command is empty");
        };
        let mut full_args = args.iter().map(String::as_str).collect::<Vec<_>>();
        full_args.push(&path_string);
        let code = run_local_inherit(program, &full_args)?;
        if code != 0 {
            bail!("editor exited with code {code}");
        }
        return Ok(());
    }

    if cfg!(target_os = "macos") {
        let result = run_local("open", &["-t", &path_string], None)?;
        if result.exit_code != 0 {
            bail!("{}", result.stderr.trim());
        }
        println!("opened {path_string}");
    } else {
        println!("{path_string}");
    }
    Ok(())
}

